src/Raven.Server/Documents/DatabasesLandlord.cs

```diff
@@ -341,9 +341,10 @@ public bool ShouldDeleteDatabase(TransactionOperationContext context, string dbN
 
             // We materialize the values here because we close the read transaction
             var record = rawRecord.MaterializedRecord;
-            context.CloseTransaction();
 
-            DeleteDatabase(dbName, deletionInProgress, record);
+            context.Transaction.InnerTransaction.LowLevelTransaction.OnDispose += _ =>
```

Comment:
We close the transaction here so it will not be opened while actually deleting the database (can take a long time).
In shard database, the raw database record of the shard database is built from the main raw database record.
On closing the transaction here we also disposing the main raw database record so in the next iteration we read from a freed memory.
I think the fact we close the transaction here is surprising while we call this function from inside scope of `using` of the transaction.
So I think it would be more clear to keep the transaction closing on the closing of the `using` scope and just register the delete with the relevant data.
The only thing that bothers me is the name of the event `OnDispose` even though it called in the end (and actually after) of transaction dispose so the transaction is already closed while doing the actual delete.

Reply:
@arekpalinski can you verify this?

Reply:
LGTM. Yes, `OnDispose` is called after we return the write transaction lock. @haludi do you think we should backport this change to 5.2?

Reply:
It doesn't harm anything right now in a regular database but I think it makes much more sense.
It really surprising to find that the transaction was closed in inner function while its `using` scope is still "active"
So in the bottom line, yes I think we should fix it on v5.2 as well